File: 04 - Largest Palindrome Product/python/main.py
import math

def reverseNumber(num):
    reversed = 0
    while (num > 0):
        reversed = (reversed * 10) + (num % 10)
        num = math.floor(num / 10)
    return reversed

def isPalindrome(num):
    return num == reverseNumber(num)

def largestPalindromeProduct(low, high):
    palindromes = []
    for i in range (low, high):
        for j in range (low, high):
            product = i * j
            if isPalindrome(product):
                palindromes.append(product)
    palindromes.sort()
    return palindromes[-1]

result = largestPalindromeProduct(100, 1000)
print(result)

File: 04 - Largest Palindrome Product/java/src/Main.java
import java.util.ArrayList;
import java.util.Collections;

public class Main {
    public static void main(String[] args) {
        int result = largestPalindromeProduct(100, 1000);
        System.out.println("result: " + result);
    }

    static int reverseNumber(int num) {
        int reversed = 0;
        while (num > 0) {
            reversed = (reversed * 10) + (num % 10);
            num = num / 10;
        }
        return reversed;
    }

    static boolean isPalindrome(int num) {
        boolean result = num == reverseNumber(num);
        return result;
    }

    static int largestPalindromeProduct(int low, int high) {
        ArrayList<Integer> palindromes = new ArrayList<Integer>();
        for (int i = low; i < high; i++) {
            for (int j = low; j < high; j++) {
                int product = i * j;
                if (isPalindrome(product)) {
                    palindromes.add(product);
                }
            }
        }
        //palindromes.sort();
        Collections.sort(palindromes);
        return palindromes.get(palindromes.size() - 1);
    }
}
